modules/ip6/datapath/ip6_error.c

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2024 Christophe Fontaine
+
+#include <gr_datapath.h>
+#include <gr_graph.h>
+#include <gr_icmp6.h>
+#include <gr_ip6_control.h>
+#include <gr_ip6_datapath.h>
+#include <gr_log.h>
+#include <gr_mbuf.h>
+
+#include <rte_common.h>
+#include <rte_graph_worker.h>
+#include <rte_ip.h>
+
+enum edges {
+	ICMP_OUTPUT = 0,
+	NO_HEADROOM,
+	NO_IP,
+	EDGE_COUNT,
+};
+
+static uint16_t
+ip6_error_process(struct rte_graph *graph, struct rte_node *node, void **objs, uint16_t nb_objs) {
+	struct icmp6_err_dest_unreach *du;
+	struct icmp6_err_ttl_exceeded *te;
+	struct ip6_local_mbuf_data *d;
+	const struct iface *iface;
+	struct rte_ipv6_hdr *ip;
+	icmp6_type_t icmp_type;
+	struct rte_mbuf *mbuf;
+	struct nexthop6 *nh;
+	struct icmp6 *icmp6;
+	rte_edge_t edge;
+
+	icmp_type = node->ctx[0];
+
+	for (uint16_t i = 0; i < nb_objs; i++) {
+		mbuf = objs[i];
+
+		// Get the pointer to the start of the ipv6 header before
+		// prepending any data
+		ip = rte_pktmbuf_mtod(mbuf, struct rte_ipv6_hdr *);
+
+		// https://www.rfc-editor.org/rfc/rfc4443.html#section-3
+		// ICMPv6 error messages should contain "As much of invoking
+		// packet as possible without the ICMPv6 packet exceeding the
+		// minimum IPv6 MTU (1280)"
+		rte_pktmbuf_trim(mbuf, RTE_MAX(0, (int)rte_pktmbuf_data_len(mbuf) - IP6_MIN_MTU));
```

Comment:
data_len is the size of the segment.
You want pkt_len.

Reply:
No point in calling rte_pktmbuf_trim with a 0 length.

Reply:
Ok, I'll add a conditional. data_len only works if there is a single segment?